sql/sql_select.cc

```diff
@@ -946,12 +949,14 @@ int SELECT_LEX::vers_setup_conds(THD *thd, TABLE_LIST *tables, COND **where_expr
           max_time.second_part= TIME_MAX_SECOND_PART;
           curr= newx Item_datetime_literal(thd, &max_time,
                                             TIME_SECOND_PART_DIGITS);
-          cond1= newx Item_func_eq(thd, row_end, curr);
+          cond1= make_expr_or_is_not_null(
+              thd, newx Item_func_eq(thd, row_end, curr), row_end);
         }
         else
         {
           curr= newx Item_int(thd, ULONGLONG_MAX);
-          cond1= newx Item_func_eq(thd, row_end, curr);
+          cond1= make_expr_or_is_not_null(
+              thd, newx Item_func_eq(thd, row_end, curr), row_end);
```

Comment:
Remove these changes, fix this code below:
```c++
       if (table->is_view_or_derived())
         vers_cond= or_items(thd, vers_cond, newx Item_func_isnull(thd, row_end));
```

Reply:
I've tried this
```diff
diff --git a/sql/sql_select.cc b/sql/sql_select.cc
index 0b07e24ce56..1aef575dfb8 100644
--- a/sql/sql_select.cc
+++ b/sql/sql_select.cc
@@ -949,14 +949,12 @@ int SELECT_LEX::vers_setup_conds(THD *thd, TABLE_LIST *tables, COND **where_expr
           max_time.second_part= TIME_MAX_SECOND_PART;
           curr= newx Item_datetime_literal(thd, &max_time,
                                             TIME_SECOND_PART_DIGITS);
-          cond1= make_expr_or_is_not_null(
-              thd, newx Item_func_eq(thd, row_end, curr), row_end);
+          cond1= newx Item_func_eq(thd, row_end, curr);
         }
         else
         {
           curr= newx Item_int(thd, ULONGLONG_MAX);
-          cond1= make_expr_or_is_not_null(
-              thd, newx Item_func_eq(thd, row_end, curr), row_end);
+          cond1= newx Item_func_eq(thd, row_end, curr);
         }
         break;
       case SYSTEM_TIME_AS_OF:
@@ -1035,7 +1033,8 @@ int SELECT_LEX::vers_setup_conds(THD *thd, TABLE_LIST *tables, COND **where_expr
         and_items(thd,
           cond2,
           cond1));
-      if (table->is_view_or_derived())
+      if (table->is_view_or_derived() ||
+          vers_conditions.type == SYSTEM_TIME_UNSPECIFIED)
         vers_cond= or_items(thd, vers_cond, newx Item_func_isnull(thd, row_end));
     }
   } // for (table= tables; ...)
```

It seems to be incorrect as it appends to the whole expression not to the `row_end == MAX`:
```sql
where (`t1`.`row_end` = MAX_RESULT or `t1`.`row_end` is null) and (`t2`.`row_end` = MAX_RESULT or `t2`.`row_end` is null)
where (`t1`.`row_end` = MAX_RESULT or `t1`.`row_end` is null) and `t2`.`row_end` = MAX_RESULT or `t2`.`row_end` is null
```

Reply:
Keep your original condition. Replace `make_expr_or_is_not_null()` with `or_items()` and merge.

Reply:
In case your condition is added, don't add it second time. Use something like this:

```
 if (table->is_view_or_derived() && vers_conditions != SYSTEM_TIME_UNSPECIFIED)
         vers_cond= or_items(thd, vers_cond, newx Item_func_isnull(thd, row_end));
```